### PR TITLE
Implement String(contentsOf:) family in Swift

### PR DESCRIPTION
**Pull request by parkera**
Now that we have all of the underlying pieces in place for data reading/writing and string encoding - finish the work to unify the implementations of `String(contentsOf:)`.

### CODE REVIEW

**Comment by parkera** [comment]
@swift-ci test

**Comment by parkera** [comment]
@swift-ci test windows

File: Benchmarks/Benchmarks/String/BenchmarkString.swift [hunk 1]
```diff
@@ -106,4 +106,84 @@ let benchmarks = {
         
     // MARK: - UTF32
 
+    Benchmark("utf32-encode", configuration: .init(warmupIterations: 1, scalingFactor: .kilo)) { benchmark in
```
Comment (parkera) [inline Benchmarks/Benchmarks/String/BenchmarkString.swift:109]: These new benchmarks round out the UTF16/8 ones from the previous String work.

File: Sources/FoundationEssentials/Data/Data+Reading.swift [hunk 2]
```diff
@@ -25,24 +25,34 @@ import CRT
...
 private func readExtendedAttributesFromFileDescriptor(_ fd: Int32, attrsToRead: [String]) -> [String : Data] {
-#if canImport(Darwin) && !NO_FILESYSTEM
+#if !NO_FILESYSTEM
     var output: [String : Data] = [:]
     for key in attrsToRead {
         key.withCString { keyStr in
             let maxXAttrLength = 1000
             withUnsafeTemporaryAllocation(of: CUnsignedChar.self, capacity: maxXAttrLength) { buf in
-                let result = fgetxattr(fd, keyStr, buf.baseAddress, maxXAttrLength, 0, 0)
+                let result = _fgetxattr(fd, keyStr, buf.baseAddress, maxXAttrLength, 0, 0)
```
Comment (parkera) [inline Sources/FoundationEssentials/Data/Data+Reading.swift:45]: @compnerd - could use your advice here on the Windows version of this code.

File: Sources/FoundationEssentials/String/String+Essentials.swift [hunk 3]
```diff
@@ -80,169 +80,4 @@ extension String {
...
-    /// Creates a new string equivalent to the given bytes interpreted in the
-    /// specified encoding.
-    ///
-    /// - Parameters:
-    ///   - bytes: A sequence of bytes to interpret using `encoding`.
-    ///   - encoding: The encoding to use to interpret `bytes`.
-    public init?<S: Sequence>(bytes: __shared S, encoding: Encoding)
-        where S.Iterator.Element == UInt8
-    {
-        // TODO: Move init?(bytes:encoding) to Swift
```
Comment (parkera) [inline Sources/FoundationEssentials/String/String+Essentials.swift:95]: Resolved

File: Sources/FoundationEssentials/Data/Data+Reading.swift (hunk 2 above)
Comment (compnerd) [reply Sources/FoundationEssentials/Data/Data+Reading.swift:45]: Windows doesn't really have "extended attributes". The permission model on Windows is pretty rich and allows for per-identity ACLing of different forms. I believe that we currently do no really handle that per se.

Comment (parkera) [reply Sources/FoundationEssentials/Data/Data+Reading.swift:45]: Ok - the purpose of this xattr is to store some metadata about the encoding of the file. We can't guarantee its presence on non-Darwin platforms, either, so we could just modify this to do nothing on Windows if you think that's correct.

Comment (parkera) [reply Sources/FoundationEssentials/Data/Data+Reading.swift:45]: Actually, I already did that above in the stub. It looks like we're good to go for now.

**Review by jmschonfeld — APPROVED** [review]

**Pull request merged**